### PR TITLE
Deprecate [Theorem ... with].

### PR DESCRIPTION
**Pull request by Zimmi48**
**Kind:** clean-up

- [x] Corresponding documentation was added / updated

This is a proposal to deprecate (and subsequently remove the syntax) `Theorem ... with` which allows to define mutually recursive fixed points in proof-editing mode. The migration path is trivial (replace `Theorem` with `Fixpoint` which also supports not providing a body and using the proof-editing mode instead). Preliminary testing has shown that this is only used (among the projects in our CI) in Equations (cc @mattam82) and CompCert (cc @xavierleroy).

I find the command inconsistent and confusing because `Theorem` will change its meaning if there is a `with` or not (will insert a `fix` / `cofix` or not) whereas `Fixpoint` / `CoFixpoint` won't (they will always insert a `fix` / `cofix`).
There are other additional inconsistencies due to a partly different implementation. See #6567 (should the limitation of the `Fixpoint` form noted in this issue be addressed before deprecating the `Theorem` form?).

This makes it also more complicated to unite `Definition` and `Theorem` as was my intention with the project of introducing `Theorem ... := ...`. In this project, the only practical remaining difference between `Definition` and `Theorem` will be the default resulting opacity (but even this could be made tweakable using an attribute).

### CODE REVIEW

**Review by cpitclaudel — APPROVED** [review]
The doc changes look good to me (I don't have an opinion on the code change).

**Comment by Zimmi48** [comment]
I'd like to have this in 8.9* so can developers comment and possibly also users of the feature @xavierleroy (for CompCert) and @mattam82 (for Equations)?

*If this eventually get merged, I'd like to have it merged rather sooner than later so that we can start benefiting from this clean-up in 8.10.

**Comment by Zimmi48** [comment]
ping @mattam82 @herbelin @xavierleroy (if nobody protests, it probably means that it is OK to remove this syntax, doesn't it?)

**Comment by herbelin** [comment]
> ping @mattam82 @herbelin @xavierleroy (if nobody protests, it probably means that it is OK to remove this syntax, doesn't it?)

I have mixed feelings. In some sense, `Theorem with` is often deceiving and thus not to recommend but the reason is that it is difficult to control the guard in interactive mode and one easily arrives to a proof which we believe is guarded but which, e.g. due to some rewriting in the way, fails to be recognized as guarded.

In itself, I don't see it as a problem of `Theorem with`, but as a weakness of the guard checker. In particular, interactive development of `Fixpoint` suffers also of this weakness.

About the unification between `Definition` and `Theorem`, maybe the young generation is so much fed of proofs-as-programs correspondence that they feel that it is an error not to unify the syntax of `Definition` and `Theorem`. For me, they remain different and, in particular, I don't see `Theorem with` as bad, but maybe I'm too old in my way of thinking :smile:

**Comment by xavierleroy** [comment]
First, I think changes to the vernacular language, especially removal of constructs that are actively used in one of your flagship projects, need to be justified better than just "I find it confusing", need to be discussed widely (e.g. on coq-dev), and need consensus beyond @Zimmi48 and @herbelin.

Second, for a very long time Coq sucked at proofs by mutual induction.  The `Scheme...with` dance is awful, `Combined Scheme` (introduced by @mattam82 after I complained) is only slightly better.  `Theorem...with` is perhaps a bit temperamental but opens new, better ways to conduct those proofs by mutual induction.  This should be encouraged and developed further rather than removed.

Third, I did not know that `Fixpoint ... with` can be used in conjunction with interactive proof mode.  That may be obvious because "it's just like `Definition`", but it is not intuitive to me so I would not have guessed.  @mattam82: did you know about this when you introduced the `Theorem...with` syntax?

Fourth, I agree with @herbelin that for presentation purposes it helps to write `Theorem` for things that are theorems and `Definition` or `Fixpoint` for things that are definitions.  I am familiar with Curry-Howard, thank you very much, but the people who read and review the CompCert development (e.g. certification authorities) may not, and I see no reason to confuse them further.

So, in summary: you need to build a stronger case for the removal of this feature and offer better alternatives than `Fixpoint...with` in proof mode.

**Comment by Zimmi48** [comment]
Thank you very much for this reply. No need to get anxious: if I pinged you three times to get your opinion, it's because I thought I needed it before proceeding. I'll close this for now since the goal was to get it in 8.9 but longer term, would an alternative like `Inductive Theorem ... with` (which would also have the advantage of supporting `Inductive Theorem` for non-mutual cases; possibly with the attribute syntax for the `Inductive` part), be a good solution to you?
